library/core/tests/async_iter/mod.rs

```diff
@@ -1,17 +1,18 @@
 use core::async_iter::{self, AsyncIterator, IntoAsyncIterator};
+use core::future::Future;
 use core::pin::pin;
 use core::task::Poll;
 
 #[test]
 fn into_async_iter() {
     let async_iter = async_iter::from_iter(0..3);
-    let mut async_iter = pin!(async_iter.into_async_iter());
+    let mut async_iter = async_iter.into_async_iter();
 
     let waker = core::task::Waker::noop();
     let mut cx = &mut core::task::Context::from_waker(&waker);
 
-    assert_eq!(async_iter.as_mut().poll_next(&mut cx), Poll::Ready(Some(0)));
-    assert_eq!(async_iter.as_mut().poll_next(&mut cx), Poll::Ready(Some(1)));
-    assert_eq!(async_iter.as_mut().poll_next(&mut cx), Poll::Ready(Some(2)));
-    assert_eq!(async_iter.as_mut().poll_next(&mut cx), Poll::Ready(None));
+    assert_eq!(pin!(async_iter.next()).poll(&mut cx), Poll::Ready(Some(0)));
+    assert_eq!(pin!(async_iter.next()).poll(&mut cx), Poll::Ready(Some(1)));
+    assert_eq!(pin!(async_iter.next()).poll(&mut cx), Poll::Ready(Some(2)));
+    assert_eq!(pin!(async_iter.next()).poll(&mut cx), Poll::Ready(None));
```

Comment:
I find this diff does a pretty good job showing the differences in pinning between the two formulations. Just like with the non-async `Iterator` trait, the AFIT-based `AsyncIterator` trait doesn't require the iterator to be pinned. Instead it's only the individual futures which have to be pinned. While the `Stream` trait requires the entire iterator to be pinned for the entire duration.

One of the things we want to investigate is the precise lowering of `async gen` blocks and functions to the AFIT-based `AsyncIterator` trait impls. The theory is that both traits should have equivalent lowerings, modulo `Stream` being able to hold local borrows across `yield` points. This is something that is currently also not allowed in non-async `gen` blocks and functions, so by also not allowing it `AsyncIterator` is merely consistent.

Reply:
> This is something that is currently also not allowed in non-async gen blocks and functions, so by also not allowing it AsyncIterator is merely consistent.

I don't think this is "merely consistent" -- I consider it to be a really big mistake of the `Iterator` trait and a major ergonomic issue that is not present in `async {}` blocks. If we can avoid this for `async gen` from the get-go, I have no idea why we wouldn't just do so.

Reply:
> I don't think this is "merely consistent" -- I consider it to be a really big mistake of the `Iterator` trait and a major ergonomic issue that is not present in `async {}` blocks. If we can avoid this for `async gen` from the get-go, I have no idea why we wouldn't just do so.


I cover some of the rationale for that in [my blog post](https://blog.yoshuawuyts.com/async-iterator-trait/#self-referential-iterators). Tldr: it seems bad to  special-case this just for async Rust; we should instead treat pinned iteration as an orthogonal concern to solve consistently for both sync and async Rust. If we solve this for async Rust in one way, we'll then need to solve it for non-async Rust a different way. And that's just one of the issues: iteration also intersects with fallibility, lending, concurrency, and parallelism. I believe we should want to approach all of those in orthogonal ways, so that the resulting solutions end up composing. That way users don't end up having to choose between outdated base traits, and unmanageable uber traits.

However we've been having these conversations within WG Async for several months now, and I don't think we're any closer to convincing each other. That's why I'm filing this PR: **I think it would be more productive if we could actually compare approaches**. Though I don't believe a `poll_next`-based trait for async iteration holds up favorably, I do think the best way to make that case is by actually being able to compare both implementations side by side.